### ISSUE TITLE
Support staging the current hunk

### ISSUE THREAD

**Issue by zw963** [dgutov/diff-hl#71]
Is there any plan to support stash operation on hunk? (as git-gutter)

Thanks.

**Comment by dgutov** [comment]
Probably not until VC supports that (i.e. not anytime soon).

**Comment by zw963** [comment]
Only support git is a easily start?

I think most of us (could access here) is use git.

**Comment by dgutov** [comment]
Feel free to send a patch to Emacs's VC implementing this operation just for Git.

**Comment by zw963** [comment]
git-gutter current exist few bug, and not worked well with linum-mode
git-gutter-fringe is less function than git-gutter, and less active development.

diff-hl is very well for me, only one things, i hope to stash code interactively
as of `git add -ip`.

**Comment by dgutov** [comment]
Your best bet for this feature is probably to get it into [commit-patch](https://github.com/caldwell/commit-patch), or a project like it. `diff-hl` lets you jump into the current hunk in the diff quickly, so the workflow should be smooth enough.

**Comment by dgutov** [comment]
OK well, a "stage hunk" command shouldn't be too hard to do, but then what? Do we change the display after that somehow? (Related: https://github.com/dgutov/diff-hl/issues/34).

**Comment by zw963** [comment]
git gutter seem like only remove left margin after add into stage.

i think that behavior work fine.

after all, In most case, we hope use `Stage hunk` to selective add some content into stage.
if all done, I will use others foreign tools to see changed in INDEX.

make selective add changed content into INDEX, i think is the key feature we need.

**Comment by valignatev** [comment]
I agree that hiding signs for staged hunk are ok, all other packages for all the editors do this, so most folks expecting this.

**Comment by dgutov** [comment]
@valignatev That will be in #161.

**Comment by zw963** [comment]
So, when we can test this feature in master branch, still not see any command for stage current hunk anyway.

**Comment by dgutov** [comment]
@zw963 Sorry, that was still only about the indicators.

I'm not sure what UI would be best for staging, and how to indicate success, for example, if the indicators still include staged changes.

**Comment by dgutov** [comment]
> Though, i consider the hunk granularity is a little not so fine enough.

Should be fixed now. Check out the new version of this command I have just pushed to master.

> (i use undo-tree, undo-tree raise error)

This discussion might be relevant: https://www.reddit.com/r/emacs/comments/qicb0z/how_to_undo_the_undo/hiiowu8/?utm_source=reddit&utm_medium=web2x&context=3

**Comment by zw963** [comment]
I am driving, i test this feature quickly, it seem like not work when i try to staging one hunk.

i only get a error message on minibuffer:

diff-hl-stage-current-hunk: Only Git supports staging; this file is controlled by git

And more,  i can't see the staging option there too.

![image](https://user-images.githubusercontent.com/549126/139582219-f4b7e6fb-ebd2-464d-8df1-e17a89fa656a.png)

**Comment by dgutov** [comment]
> diff-hl-stage-current-hunk: Only Git supports staging; this file is controlled by git

What's your value of `vc-handled-backends`? By default, it contains `Git` uppercase.

> And more, i can't see the staging option there too.

Yep, it's not there (yet?). Maybe I'll add it conditionally based on the value of `diff-hl-show-staged-changes`. To avoid cluttering the UI for those users who don't use this workflow.

**Comment by zw963** [comment]
> What's your value of `vc-handled-backends`? By default, it contains `Git` uppercase.

```
vc-handled-backends is a variable defined in vc-hooks.el.gz.

Value
(git RCS CVS SVN SCCS SRC Bzr Git Hg)

Original Value
(RCS CVS SVN SCCS SRC Bzr Git Hg Mtn)
```

I guess it changed by git-emacs package, i use several very basically feature from it.

![image](https://user-images.githubusercontent.com/549126/139584466-bd7f10e3-4af2-41f8-aef2-f84440cf5512.png)

After comment it, it start to working now. thank you.




> Yep, it's not there (yet?). Maybe I'll add it conditionally based on the value of `diff-hl-show-staged-changes`. To avoid cluttering the UI for those users who don't use this workflow.

Okay, thank you, `stage-current-hunk` is very useful, at least for me, e.g. 

following is my process:

1. i do many changes in current buffer.
2. it is time to create a commit and push code, but not all feature is done, i only want to add several of thems.
3. select which hunk need to be stage visually.
4. do git commit on changes in git stage.
5. if i want to do it again, i can always use  `git reset`

**Comment by zw963** [comment]
> This discussion might be relevant: https://www.reddit.com/r/emacs/comments/qicb0z/how_to_undo_the_undo/hiiowu8/?utm_source=reddit&utm_medium=web2x&context=3

Not sure , most of time, i just care about the undo history which before restart/close current emacs session, i use tree-macs for years, the issue is, once undo-tree history can't move back because some error raised on minibuffer occasionally, but this
seem like not happen recent years, but anyway, this cause me not so trust undo-tree, though, the new format maybe resolve this issue.

**Comment by yugaego** [comment]
Thank you for adding this feature.

> Yep, it's not there (yet?). Maybe I'll add it conditionally based on the value of diff-hl-show-staged-changes. To avoid cluttering the UI for those users who don't use this workflow.

I think it may be useful, because it would:
- be easier to discover this command is available,
- correspond to other (when it makes sense) commands, providing the default keybindings also for the "show" state.

(And looks like it has a potential to match my workflow nicely too.)

**Comment by dgutov** [comment]
> I guess it changed by git-emacs package, i use several very basically feature from it.

Is it popular? I guess I could change the check to be case-insensitive.

> if i want to do it again, i can always use git reset

Should we add a dedicated command `diff-hl-unstage-file`?

**Comment by dgutov** [comment]
> be easier to discover this command is available,

It's a good reason to have it.

Now added. The issue of mismatching bindings between the UIs is still there, though.

**Comment by zw963** [comment]
> Is it popular? I guess I could change the check to be case-insensitive.

No, it never update since 5 years, at that time, i pick it for very limit usage. 

e.g. i only use git-emacs's git-log-view-mode and other useful mode,  and in that model, i can use git-emacs give me some useful command for get the git hash under the cursor point.

![image](https://user-images.githubusercontent.com/549126/139624135-950bff06-8d31-40ba-ba61-a123bd01a3c5.png)

Then, i write some function for run beyond-compare for check diff info, e.g. when i pressing `=` on above cursor. will see
bc4 running like this.

![image](https://user-images.githubusercontent.com/549126/139624528-b0798f01-be91-4683-b578-dc55760108d6.png)

you can check my code [here](https://github.com/zw963/.emacs.d/blob/master/plugins/git/git-emacs/git-emacs_init.el)

anyway, i doubt those mode's feature should be available as emacs builtin as diff-hl depend on, i just not research it for now

 i found after comment `(add-to-list 'vc-handled-backends 'git)` out, my above hack not works anymore with a similar message. `git--require-buffer-in-git: Current buffer must be a file under git revision control`, but i think you don't need think about this very very old package.

> Should we add a dedicated command diff-hl-unstage-file?

Maybe useful? anyway, i always can do it use `git reset` or `git reset -- some_file`.
in fact, git-emacs package have a function name git-reset. https://github.com/zw963/.emacs.d/blob/master/plugins/git/git-emacs/git-emacs.el#L1698-L1721, but i always forget to use it on emacs. maybe because i am a command line geek, or because we could not staging hunk in emacs before, so, i have to add hunks with `git add -p`, so use `git reset` is more conveniently.

**Comment by zw963** [comment]
> > be easier to discover this command is available,
> 
> It's a good reason to have it.
> 
> Now added. The issue of mismatching bindings between the UIs is still there, though.

I get error like this.

```
Debugger entered--Lisp error: (void-variable diff-hl-show-hunk--inline-popup-map)
  byte-code("\302\10\303\304#\210\302\11\305\306#\210\302\11\307\310#\210\302\11\311\312#\210\313\314!\207" [diff-hl-command-map diff-hl-show-hunk--inline-popup-map define-key [(*)] diff-hl-show-hunk-or-show-first-hunk "p" diff-hl-show-hunk-previous "n" diff-hl-show-hunk-next "r" diff-hl-show-hunk-revert-hunk provide diff-hl_init] 4)
  #<subr require>(diff-hl_init)
  apply(#<subr require> diff-hl_init)
  require(diff-hl_init)
  load-with-code-conversion("/home/zw963/Dropbox/common/.emacs.d/early-init.el" "/home/zw963/Dropbox/common/.emacs.d/early-init.el" nil nil)
  load("/home/zw963/Dropbox/common/.emacs.d/early-init.el")
  (progn (load absolute-path))
  (if (file-exists-p absolute-path) (progn (load absolute-path)))
  (let ((absolute-path (expand-file-name file-relative-path (file-name-directory (or load-file-name buffer-file-name))))) (if (file-exists-p absolute-path) (progn (load absolute-path))))
  relative-load("Dropbox/common/.emacs.d/early-init.el")
  load-with-code-conversion("/home/zw963/.emacs" "/home/zw963/.emacs" t t)
  load("~/.emacs" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0x15485576a0a666fe>) #f(compiled-function () #<bytecode -0x1f3c686ddc0cc375>) t)
  command-line()
  normal-top-level()
```

my config: https://github.com/zw963/.emacs.d/blob/master/plugins/diff-hl/diff-hl_init.el

**Comment by dgutov** [comment]
> I get error like this.

You'll need to update the references in your config to use `diff-hl-show-hunk-map`.

**Comment by dgutov** [comment]
> No, it never update since 5 years, at that time, i pick it for very limit usage.

All right, then. Its repository is on github, and if someone is using it in earnest, they can pick up the development.

BTW, I think setting `diff-hl-show-staged-changes` to nil probably didn't work for you as well before commenting out that line, since the related code also compares the file's backend to `Git`.

> Maybe useful? anyway, i always can do it use git reset or git reset -- some_file.

Me too, but a "symmetric" command makes sense to have anyway, even if people use it only occasionally.

Added `diff-hl-unstage-file`. And with that, it seems this issue can be closed.

**Comment by zw963** [comment]
Thanks,  diff-hl-unstage-file works for me.

> You'll need to update the references in your config to use diff-hl-show-hunk-map.

Yes, it works now, i can see `(S)Stage option` when do `diff-hl-show-hunk`.

There i consider still have a improve, let me try to explain.

Staging is a no dangerous operation, so, it maybe useful do staging successively.

But, when i pressing 'S' on following view, it exit `diff-hl-show-hunk` and, i have to run it again to check next hunk, so, when i want to add  3 hunks, i have to:

1.  run diff-hl-show-hunk
2.  pressing S
3. run diff-hl-show-hunk again
4. pressing S
5. run diff-hl-show-hunk again
6. Pressing S

so, if we consider add a option, when use do staging from `diff-hl-show-hunk`, after user pressing 'S', we can switch to next hunk automatically, that will save a lot time.

------------------------

In fact, when i try to do above sequence process, i can always get following error message when try to do staging the second time

```
Running in foreground: git --no-pager ls-files -s diff-hl.el
Done (status=0): git --no-pager ls-files -s diff-hl.el
Running in foreground: git --no-pager ls-files -s diff-hl.el
Done (status=0): git --no-pager ls-files -s diff-hl.el
Running in foreground: git --no-pager apply --cached ../../../../../../../tmp/diff-hl-stage-patchwFNKSP
Done (status=0): git --no-pager apply --cached ../../../../../../../tmp/diff-hl-stage-patchwFNKSP
Hunk staged
Running in foreground: git --no-pager ls-files -s diff-hl.el
Done (status=0): git --no-pager ls-files -s diff-hl.el
Running in foreground: git --no-pager cat-file blob 1b0c946eef3bf685acd4b956366c0c9afd046f39 .
Done (status=0): git --no-pager cat-file blob 1b0c946eef3bf685acd4b956366c0c9afd046f39 .
Running in foreground: git --no-pager ls-files -s diff-hl.el
Done (status=0): git --no-pager ls-files -s diff-hl.el
Running in foreground: git --no-pager ls-files -s diff-hl.el
Done (status=0): git --no-pager ls-files -s diff-hl.el
Running in foreground: git --no-pager apply --cached ../../../../../../../tmp/diff-hl-stage-patchkBTX1x
vc-do-command: Failed (status 128): git --no-pager apply --cached ../../../../../../../tmp/diff-hl-stage-patchkBTX1x
```

And, then, diff-hl seem like live in a strange status.

i can see fridge show exists a hunk under current point, but when run `diff-hl-show-hunk`, no any UI changes show under current hunk.

Following is the backtrace

```
Debugger entered--Lisp error: (error "Failed (status 128): git --no-pager apply --cached...")
  error("Failed (%s): %s" "status 128" "git --no-pager apply --cached ../../../../../../.....")
  vc-do-command(#<buffer  *string-output*> 0 "git" "/tmp/diff-hl-stage-patcholZQjg" "--no-pager" "apply" "--cached")
  vc-git-command(#<buffer  *string-output*> 0 "/tmp/diff-hl-stage-patcholZQjg" "apply" "--cached")
  diff-hl-stage-current-hunk()
  diff-hl-show-hunk-stage-hunk()
  funcall-interactively(diff-hl-show-hunk-stage-hunk)
  command-execute(diff-hl-show-hunk-stage-hunk)
```

Thank you.

**Comment by dgutov** [comment]
> so, if we consider add a option, when use do staging from diff-hl-show-hunk, after user pressing 'S', we can switch to next hunk automatically, that will save a lot time.

First of all, you can do staging without "show hunk", it's just `C-x v S`, and then go to the next one with `]`. But I can add a slightly different command which would switch too.

> i can see fridge show exists a hunk under current point, but when run diff-hl-show-hunk, no any UI changes show under current hunk

Did you see the instructions in the message after `diff-hl-stage-current-hunk` completes successfully?

**Comment by dgutov** [comment]
> Sorry, what you mean is, C-x v S can staging current hunk, but, i have to pressing C-x v ], right? it is conveniently, though, not so used to it.

If you have `smartrep` installed, or enable `repeat-mode` from Emacs 28, you can shorten it to `C-x v S ] S ] S ] ]`.

> Following is message buffer info:

OK. Questions:

1. Does the staging actually succeed? Was the hunk added correctly? You can check with `git diff --cached`.
2. What if your value of `diff-hl-show-staged-changes`?
3. What does `(vc-backend buffer-file-name)` evaluate to?

**Comment by zw963** [comment]
> If you have `smartrep` installed, or enable `repeat-mode` from Emacs 28, you can shorten it to `C-x v S ] S ] S ] ]`.

Thanks for let me know, repeat-mode works, it awesome

And, staging failing issue was fixed since yesterday commit.

https://github.com/dgutov/diff-hl/compare/c9dbc89..47f8724?w=1

BTW: i set diff-hl-show-staged-changes to nil, backend is Git.

**Comment by dgutov** [comment]
> And, staging failing issue was fixed since yesterday commit

Wasn't supposed to fix anything like that, but no matter, glad it's working for you.

> BTW: i set diff-hl-show-staged-changes to nil, backend is Git.

:+1:

**Comment by zw963** [comment]
> Wasn't supposed to fix anything like that, but no matter, glad it's working for you.

But i can reproduce same issue after revert to c9dbc89. :smile:

**Comment by irisjae** [comment]
I'm facing problems staging as well. Debugging seems to indicate this occurs due to `error: corrupt patch at line ...`, which is probably caused by whitespace issues.

**Comment by dgutov** [comment]
@irisjae It would be great if you could file a separate issue, with steps for reproduction.

And your OS and version of Emacs.

**Comment by irisjae** [comment]
Okay, good suggestion. I'll file this up once i have a bit more time.
